src/peft/tuners/reft/layer.py

```diff
@@ -0,0 +1,195 @@
+            if weight.dtype.is_floating_point or weight.dtype.is_complex:
+                self.to(weight.device, dtype=weight.dtype)
+            else:
+                self.to(weight.device)
+        self.set_adapter(self.active_adapters)
+
+    def get_delta_weight(self, adapter_name: str) -> torch.Tensor:
+        raise NotImplementedError
+
+    def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+        previous_dtype = x.dtype
+
+        if self.disable_adapters:
+            if self.merged:
+                self.unmerge()
+            output = self.base_layer(x, *args, **kwargs)
+        elif self.merged:
+            output = self.base_layer(x, *args, **kwargs)
+        else:
+            result = self.base_layer(x, *args, **kwargs)
+
+            # Execute all the adapters
+            for active_adapter in self.active_adapters:
+                if active_adapter not in self._available_adapters:
+                    continue
+
+                rotate_layer = self.reft_R[active_adapter]
+                learned_source = self.reft_A[active_adapter]
+                dropout = self.dropout[active_adapter]
+
+                result = result.to(torch.float32)
+                if self.first_n[active_adapter] == 0 and self.last_n[active_adapter] == 0:
+                    rotated_base = rotate_layer(result)
+                    offset = torch.matmul((learned_source(result) - rotated_base), rotate_layer.weight)
+                    output = result + offset
+                else:
+                    first_n = self.first_n[active_adapter]
+                    last_n = self.last_n[active_adapter]
+                    loc = torch.cat([torch.arange(first_n), torch.arange(result.shape[1]-last_n, result.shape[1])])
+                    selected_results = torch.gather(result, 1, loc)
+                    rotated_base = rotate_layer(selected_results)
+                    offset = torch.matmul((learned_source(selected_results) - rotated_base), rotate_layer.weight)
+                    output.scatter_(1, loc, offset)
+                output = dropout(output)
+        output = output.to(previous_dtype)
+        return output
+
+
+class Linear(LoReftLayer):
+    """Reft implemented in Linear layer"""
+
+    def __init__(
+        self,
+        base_layer: nn.Module,
+        adapter_name: str = "default",
+        r: int = 0,
+        alpha: float = 0.0,
+        loc: str = None,
+        dropout: float = 0.0,
+        **kwargs,
+    ):
+        super().__init__(base_layer)
+
+        # Create adapter and set it active
+        self._active_adapter = adapter_name
+        self.update_layer(adapter_name, r, alpha, loc, dropout, **kwargs)
+
+    def _get_delta_activations(
+        self, adapter_name: str, input: torch.Tensor, *args: Any, **kwargs: Any
+    ) -> torch.Tensor:
+        raise NotImplementedError
+
+
+    def __repr__(self) -> str:
+        rep = super().__repr__()
+        return "reft." + rep
```

Comment:
```suggestion
        return "loreft." + rep
```

In case we add more ReFT layers in the future. This is not consistent with the module name, but I think that's not crucial.